pkg/metrics/metrics_test.go

```diff
@@ -0,0 +1,264 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"math"
+	"net/http"
+	"os"
+	"reflect"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	cmg "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+var (
+	mgr MetricsManager
+	srv *http.Server
+	wg  *sync.WaitGroup
+)
+
+const (
+	httpPattern = "/metrics"
+	addr        = "localhost:12134"
+	httpAddr    = "http://" + addr + httpPattern
+)
+
+func initMgr() {
+	wg = &sync.WaitGroup{}
+	wg.Add(1)
+	mgr = NewMetricsManager()
+	var err error
+	if srv, err = mgr.StartServing(httpPattern, addr, nil, wg); err != nil {
+		log.Fatalf("failed to start serving [%v]", err)
+	}
+}
+
+func shutdown() {
+	if err := srv.Shutdown(context.Background()); err != nil {
+		panic(err)
+	}
+	wg.Wait()
+}
+
+func TestMain(m *testing.M) {
+	initMgr()
+	c := m.Run()
+	shutdown()
+	os.Exit(c)
+}
+
+func TestNew(t *testing.T) {
+	if mgr == nil {
+		t.Errorf("failed testing new")
+	}
+}
+
+func TestStartServing(t *testing.T) {
+	if _, err := http.Get(httpAddr); err != nil {
+		t.Errorf("failed to get response from server %v", err)
+	}
+}
+
+func TestDropOperation(t *testing.T) {
+	op := Operation{
+		Name:       "should-have-been-dropped",
+		Driver:     "driver",
+		ResourceID: types.UID("uid"),
+	}
+	mgr.OperationStart(op)
+	mgr.DropOperation(op)
+	rsp, err := http.Get(httpAddr)
+	if err != nil || rsp.StatusCode != http.StatusOK {
+		t.Errorf("failed to get response from server %v, %v", err, rsp)
+	}
+	r, err := ioutil.ReadAll(rsp.Body)
+	if err != nil {
+		t.Errorf("failed to read response body %v", err)
+	}
+	if strings.Contains(string(r), op.Name) {
+		t.Errorf("found metric should have been dropped for operation [%s] [%s]", op.Name, string(r))
+	}
+	// re-add
+	mgr.OperationStart(op)
+	time.Sleep(300 * time.Millisecond)
+	mgr.RecordMetrics(op, Success)
+	expected :=
+		`# HELP snapshot_controller_operation_total_seconds [ALPHA] Total number of seconds spent by the controller on an operation from end to end
```

Comment:
What does "ALPHA" mean?

Reply:
This is a fixed format from k8s.io/component-base/metrics
when initializing a metrics (in this case a histogram), this flag shows the stability level of this metrics, and by default its ALPHA.
see [here](https://github.com/kubernetes/kubernetes/blob/d56aaf77b929f70e740aab9c965a9b932187d036/staging/src/k8s.io/component-base/metrics/opts.go#L57)